<a href="https://www.kaggle.com/code/alinahurtova/practice-1-nn?scriptVersionId=353876066" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

Імпорт модулів

In [277]:
import torch

# Завдання 1

Перегляньте інформацію про доступні процесори, та виведіть про них інформацію

In [278]:
print(torch.cuda.is_available())  # True/False

if torch.cuda.is_available():
    print(torch.cuda.device_count())  # Кількість GPU
    print(torch.cuda.get_device_name(0))  # Назва першого GPU

tensor = torch.tensor([1, 2, 3])
print(tensor.device)  # CPU

True
2
Tesla T4
cpu


# Завдання 2
Підключіться до процесора GPU, та знову виведіть інформацію про доступні процесори

In [279]:
tensor_gpu = tensor.to("cuda")

tensor_gpu

tensor([1, 2, 3], device='cuda:0')

# Завдання 3
Створіть випадковий тензор розміру (2, 3). Виведіть інформацію про його:

* розмір
* тип даних
* процесор
* наявність градієнта

In [280]:
tensor = torch.randn(2,3)

In [281]:
tensor

tensor([[-1.5082, -0.3983, -0.1547],
        [-0.4757, -0.9796,  0.4088]])

In [282]:
tensor.dtype

torch.float32

In [283]:
tensor.shape

torch.Size([2, 3])

In [284]:
tensor.device

device(type='cpu')

# Завдання 4

Створіть 2 тензора розміру 100. Виведіть
* суму векторів
* частку векторів(перед цим замініть один елемент на 0)
* максимальне та мінімальне значення(методи .min() .max())
* середнє значення(метод .mean())
* квадрад різниці
* середнє значення квадрату різниці

In [285]:
tensor1=torch.randn(100)
tensor2=torch.randn(100)

In [286]:
tensor1

tensor([ 0.8802,  0.8835, -0.5894, -0.1571, -0.5267,  0.4530,  0.5332, -2.0385,
         0.0630,  0.0788,  0.2510, -0.1921,  0.2111, -1.8690,  0.2620,  0.5632,
         2.0596, -0.6016,  0.0461, -0.2776, -0.3256,  1.4339, -0.0304, -0.1374,
        -1.0516, -0.3622,  0.6021, -1.2897,  0.7326, -1.1931, -0.3353, -0.1809,
        -0.2321,  0.3667, -0.6441,  1.0628,  0.0452, -1.7479,  0.8866,  0.7708,
        -1.3161,  0.2740, -1.5437,  0.2489, -0.8467,  0.7119,  0.6958,  0.1003,
         0.5828, -0.2757, -1.0103, -1.5963, -1.6094, -0.3695, -1.0741,  0.5446,
         0.2007,  0.4946, -0.3526, -0.6080, -0.3617,  0.7032,  0.2469, -0.0410,
         1.4160,  0.8968, -0.0730, -0.1955,  1.0284, -1.2773, -1.6723, -0.9067,
        -0.3140,  1.0020,  0.7651,  1.1696,  1.0373, -1.5511,  0.3643, -0.2583,
         0.0476, -0.2324, -0.0693, -0.4705,  1.3742, -0.0788, -0.8319, -1.3596,
         0.9049,  0.3443,  2.3881, -0.4812,  1.3200,  0.6390, -0.3892,  2.1961,
        -0.6683,  0.8167,  0.8101, -1.00

In [287]:
tensor2

tensor([ 0.0699, -0.0630,  0.9202,  0.5690,  0.8040,  0.0456,  0.2727,  2.0711,
         0.8859,  0.6989,  0.4764,  1.9387, -1.7847,  0.2737,  1.9492,  0.8249,
        -1.5930, -0.4970,  0.4281, -0.0534, -0.0333,  0.6415,  0.8242, -0.7050,
        -1.9018, -1.5275,  0.7882,  0.1855, -0.7973,  0.5746,  1.1142,  0.0704,
         0.7015, -2.0105,  1.5619, -0.0972, -0.9081, -0.2566,  0.3036,  0.6006,
         1.0310,  0.2097, -0.5718,  1.5939,  1.6823,  0.3209, -0.0770, -0.0664,
        -0.7213,  1.3376, -1.8894, -0.3882,  0.2251, -1.0358, -0.9122,  0.3222,
        -0.0902, -1.1869,  1.3251, -0.9074,  0.5721,  0.2793, -1.9472, -1.2249,
        -0.8140, -0.3705,  0.0639, -0.7430, -0.2166,  0.1907,  1.0181, -0.4942,
         0.9452, -0.3900,  1.0417, -0.0897,  1.1166, -0.5763, -0.6045,  0.1430,
         0.8414,  0.6354,  0.2200,  0.5646, -0.6890, -0.8184, -0.0789,  0.8281,
         0.0053,  0.1578,  0.4710, -1.2386, -1.6938,  0.3076,  1.1992, -0.4996,
        -0.6221, -0.6513, -1.5416, -0.94

In [288]:
tensor2[0] = 0
tensor_div = tensor1/tensor2

In [289]:
tensor_min = tensor.min()
tensor_max = tensor.max()

In [290]:
tensor_sum = tensor1+ tensor2


In [291]:
tensor_max

tensor(0.4088)

In [292]:
tensor_min

tensor(-1.5082)

In [293]:
tensor_div

tensor([        inf, -1.4020e+01, -6.4049e-01, -2.7606e-01, -6.5514e-01,
         9.9368e+00,  1.9556e+00, -9.8426e-01,  7.1102e-02,  1.1278e-01,
         5.2699e-01, -9.9106e-02, -1.1828e-01, -6.8282e+00,  1.3442e-01,
         6.8275e-01, -1.2929e+00,  1.2104e+00,  1.0759e-01,  5.1981e+00,
         9.7799e+00,  2.2353e+00, -3.6910e-02,  1.9489e-01,  5.5294e-01,
         2.3713e-01,  7.6386e-01, -6.9526e+00, -9.1886e-01, -2.0763e+00,
        -3.0092e-01, -2.5690e+00, -3.3091e-01, -1.8237e-01, -4.1238e-01,
        -1.0934e+01, -4.9778e-02,  6.8108e+00,  2.9207e+00,  1.2833e+00,
        -1.2766e+00,  1.3066e+00,  2.6995e+00,  1.5617e-01, -5.0333e-01,
         2.2181e+00, -9.0415e+00, -1.5102e+00, -8.0799e-01, -2.0615e-01,
         5.3470e-01,  4.1123e+00, -7.1511e+00,  3.5674e-01,  1.1774e+00,
         1.6902e+00, -2.2253e+00, -4.1668e-01, -2.6611e-01,  6.7009e-01,
        -6.3224e-01,  2.5174e+00, -1.2678e-01,  3.3469e-02, -1.7396e+00,
        -2.4203e+00, -1.1422e+00,  2.6318e-01, -4.7

In [294]:
tensor_diff = tensor1-tensor2

In [295]:
tensor_diff_sq = tensor_diff**2


tensor_diff_sq

tensor([7.7472e-01, 8.9591e-01, 2.2787e+00, 5.2719e-01, 1.7707e+00, 1.6596e-01,
        6.7886e-02, 1.6889e+01, 6.7716e-01, 3.8445e-01, 5.0772e-02, 4.5407e+00,
        3.9834e+00, 4.5912e+00, 2.8467e+00, 6.8484e-02, 1.3341e+01, 1.0938e-02,
        1.4598e-01, 5.0261e-02, 8.5420e-02, 6.2793e-01, 7.3032e-01, 3.2217e-01,
        7.2288e-01, 1.3580e+00, 3.4646e-02, 2.1764e+00, 2.3405e+00, 3.1248e+00,
        2.1010e+00, 6.3192e-02, 8.7155e-01, 5.6509e+00, 4.8667e+00, 1.3456e+00,
        9.0871e-01, 2.2239e+00, 3.3991e-01, 2.8951e-02, 5.5088e+00, 4.1322e-03,
        9.4450e-01, 1.8090e+00, 6.3958e+00, 1.5283e-01, 5.9715e-01, 2.7817e-02,
        1.7008e+00, 2.6027e+00, 7.7285e-01, 1.4596e+00, 3.3652e+00, 4.4395e-01,
        2.6193e-02, 4.9458e-02, 8.4620e-02, 2.8274e+00, 2.8148e+00, 8.9612e-02,
        8.7184e-01, 1.7965e-01, 4.8137e+00, 1.4015e+00, 4.9729e+00, 1.6062e+00,
        1.8753e-02, 2.9972e-01, 1.5501e+00, 2.1548e+00, 7.2380e+00, 1.7012e-01,
        1.5857e+00, 1.9375e+00, 7.6504e-

In [296]:
tensor_diff_sq.mean()

tensor(1.9264)

In [297]:
tensor_div.mean()

tensor(inf)

# Завдання 5

У вашому розпорядженні є нейронна мережа, яка прогнозує **орієнтовну ціну квартири** на основі двох **ознак**:

* **площі квартири** (у квадратних метрах);
* **кількості кімнат**.

Вам надано готовий клас `NeuralNetwork` із методом `predict`.

1. Створіть екземпляр класу `NeuralNetwork`, передайте як параметр `device`.
2. За допомогою методу `predict()` визначте прогнозовану ціну квартири, яка:
   * має площу **65 м²**;
   * має **2 кімнати**.
     
3. Для цього створіть тензор із відповідними числами та скористайтесь методом `predict()`.
4. Усі нейромережі очікують дані розміру `(N, K)`, де:
   * `N` — кількість об'єктів, для яких виконується прогноз;
   * `K` — кількість ознак. У цьому випадку їх дві — **площа квартири та кількість кімнат**.

5. Спробуйте скористатись методом `predict()` разом із `torch.no_grad()`:

```python
with torch.no_grad():
    price = model.predict(...)
```

**Завдання:** отримайте прогнозовану ціну квартири та виведіть результат на екран.


In [298]:
import torch


class NeuralNetwork:
    def __init__(self, device="cpu"):
        self.device = device

        self.W = torch.tensor(
            [[3.3, -38.]],
            requires_grad=True,
            device=device
        )

        self.b = torch.tensor(
            [176.],
            requires_grad=True,
            device=device
        )

    def predict(self, X):
        if not isinstance(X, torch.Tensor):
            raise TypeError("X повинен бути об'єктом torch.Tensor.")

        if X.ndim != 2:
            raise ValueError(
                f"Очікується двовимірний тензор, отримано {X.shape}."
            )

        if X.shape[1] != 2:
            raise ValueError(
                f"Очікується 2 ознаки, отримано {X.shape[1]}."
            )

        if not X.is_floating_point():
            raise TypeError(
                "Тензор повинен містити числа з плаваючою комою (float32 або float64)."
            )

        if X.device != self.W.device:
            raise RuntimeError(
                f"X знаходиться на {X.device}, а модель на {self.W.device}."
            )

        return X @ self.W.T + self.b


In [299]:
# Ваш код
neural_network = NeuralNetwork(device = 'cuda')


In [300]:
tensor = torch.tensor([65, 2])

In [301]:
tensor

tensor([65,  2])

In [302]:
tensor1 = tensor.reshape(1,2)

In [303]:
tensor1.shape

torch.Size([1, 2])

In [304]:
tensor.shape

torch.Size([2])

In [305]:
tensor1 = tensor1.float()

In [306]:
tensor1 = tensor1.to('cuda')

tensor1

tensor([[65.,  2.]], device='cuda:0')

In [307]:
neural_network.predict(tensor1)

tensor([[314.5000]], device='cuda:0', grad_fn=<AddBackward0>)

In [308]:
tensor1.device

device(type='cuda', index=0)

# **Завдання 6**

У таблиці наведено інформацію про три квартири.

| **Квартира** | **Площа (м²)** | **Кількість кімнат** |
| ------------ | -------------: | -------------------: |
| Квартира 1   |             40 |                    1 |
| Квартира 2   |             60 |                    3 |
| Квартира 3   |             100 |                    2 |

Виконайте наступні кроки:

1. Створіть **три окремі тензори** (`X1`, `X2`, `X3`), кожен з яких містить дані про одну квартиру.
2. За допомогою методу `predict()` обчисліть **прогнозовану ціну** для кожної квартири окремо.
3. Створіть **один тензор** `X` розміру **(3, 2)**, де:
   * кожен рядок відповідає одній квартирі;
   * перший стовпець — площа квартири;
   * другий стовпець — кількість кімнат.

4. За допомогою одного виклику `predict(X)` отримайте **прогнозовану ціну одразу для всіх трьох квартир**.
5. Виведіть розмір тензору з трьома прогнозованими цінами, який ви отримали на кроці 4.
6. Порівняйте результати, отримані окремими викликами `predict()` та одним спільним викликом `predict(X)`.


In [309]:
# Створіть три окремі тензори (X1, X2, X3), кожен з яких містить дані про одну квартиру.

x1 = torch.tensor([40.0, 1.0])
x2 = torch.tensor([60.0, 3.0])
x3 = torch.tensor([100.0, 2.0])

In [310]:
x1 = x1.to('cuda')
x2 = x2.to('cuda')
x3 = x3.to('cuda')

x1

tensor([40.,  1.], device='cuda:0')

In [311]:
# 1 - одна квартира, 2 - кількість ознак
x1 = x1.reshape(1,2)
x2 = x2.reshape(1,2)
x3 = x3.reshape(1,2)

x1.shape

torch.Size([1, 2])

In [312]:
# За допомогою методу predict() обчисліть прогнозовану ціну для кожної квартири окремо.

neural_network.predict(x1)

tensor([[270.]], device='cuda:0', grad_fn=<AddBackward0>)

In [313]:
neural_network.predict(x2)

tensor([[260.]], device='cuda:0', grad_fn=<AddBackward0>)

In [314]:
neural_network.predict(x3)

tensor([[430.]], device='cuda:0', grad_fn=<AddBackward0>)

In [315]:

X = torch.tensor([
    [40.0, 1.0], 
    [60.0, 3.0],
    [100.0, 2.0]
])

In [316]:
X = X.to('cuda')


In [317]:
neural_network.predict(X)

tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)

# Завдання 7

Три квартири були продані, і стали відомі їхні **реальні ціни**.

| Квартира   | Реальна ціна (1000$) |
| ---------- | ---------------: |
| Квартира 1 |           210 |
| Квартира 2 |          310 |
| Квартира 3 |          360 |

Нехай прогноз моделі збережений у змінній `y_pred`, а реальні значення — у `y_true`.

Визначте, **наскільки модель помилилася**, використавши два різних підходи.

### Варіант 1

1. Знайдіть різницю між прогнозованою та реальною ціною (`y_pred - y_true`).
2. Обчисліть середнє значення цієї різниці за допомогою `torch.mean()`.

### Варіант 2

1. Знайдіть різницю між прогнозованою та реальною ціною (`y_pred - y_true`).
2. Позбудьтеся від'ємних значень за допомогою `torch.abs()`.
3. Обчисліть середнє значення отриманих помилок.

### Запитання

1. Який із двох варіантів краще характеризує якість моделі? Поясніть свою відповідь.
2. Напишіть функцію `loss_fn()`, яка реалізує обраний варіант.

```python
def loss_fn(y_pred, y_true):
    """
    Обчислює середню абсолютну помилку (Mean Absolute Error, MAE).

    Параметри:
        y_pred (torch.Tensor): прогнозовані значення моделі.
        y_true (torch.Tensor): реальні (правильні) значення.

    Повертає:
        torch.Tensor: середню абсолютну помилку.
    """
    ...
```




In [318]:
def loss_fn(y_pred, y_true):
    """
    Обчислює середню абсолютну помилку (Mean Absolute Error, MAE).

    Параметри:
        y_pred (torch.Tensor): прогнозовані значення моделі.
        y_true (torch.Tensor): реальні (правильні) значення.

    Повертає:
        torch.Tensor: середню абсолютну помилку.
    """
    MAE = y_pred - y_true
    return MAE.abs().mean()

In [319]:
y_true = torch.tensor([210.0, 310.0, 360.0])


y_true = y_true.to('cuda')

y_true

tensor([210., 310., 360.], device='cuda:0')

In [320]:
y_true = y_true.reshape(3,1)
y_true

tensor([[210.],
        [310.],
        [360.]], device='cuda:0')

In [321]:
y_pred = neural_network.predict(X)

y_pred

tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)

In [322]:
MAE = loss_fn(y_pred, y_true)

MAE

tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)

# Завдання 8

У вашому розпорядженні є вчитель, який може навчити нейронну мережу робити точніші прогнози **цін на квартири**.

Клас `Teacher` вже реалізований. Під час створення об'єкта необхідно передати модель, яку потрібно навчати:

```python
teacher = Teacher(model)
```

### Завдання

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Виведіть тензор `X`, який містить дані про квартири:
   * площу квартири;
   * кількість кімнат.

3. Виведіть тензор `y_true` з **реальними цінами квартир**.
4. Отримайте прогноз моделі `y_pred`.
5. Виведіть прогнозовані ціни квартир.
6. За допомогою власної функції `loss_fn()` обчисліть помилку моделі.
7. Виведіть значення помилки.
8. Скористайтеся методом `teach()` класу `Teacher`, передавши йому обчислену помилку.
9. Після навчання знову отримайте прогноз `y_pred`.
10. Виведіть нові прогнозовані ціни квартир.
11. Знову обчисліть та виведіть помилку.
12. Зробіть висновок: **чи стала модель прогнозувати ціни квартир точніше після одного кроку навчання?**



In [323]:
import torch


class Teacher:
    def __init__(self, model, lr=0.02):
        """
        Створює вчителя для навчання моделі.

        Параметри:
            model (NeuralNetwork): модель, яку потрібно навчати.
            lr (float): швидкість навчання.
        """
        self.model = model

        self.optimizer = torch.optim.SGD(
            [model.W, model.b],
            lr=lr
        )

    def teach(self, loss):
        """
        Виконує один крок навчання.

        Параметри:
            loss (torch.Tensor): значення функції втрат.
        """
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()


Алгоритм
X → прогноз → помилка → перенавчання моделі → backward → зміна W,b → новий прогноз → нова помилка

In [324]:
# 1. Створюємо Teacher, передаємо модель
teacher = Teacher(neural_network)

In [325]:
# 2. Виводимо дані про квартири
print("X (flats):")
print(X)

X (flats):
tensor([[ 40.,   1.],
        [ 60.,   3.],
        [100.,   2.]], device='cuda:0')


In [326]:
# 3. Реальні ціни
print("y_true:")
print(y_true)

y_true:
tensor([[210.],
        [310.],
        [360.]], device='cuda:0')


In [327]:
# 4. Прогноз моделі до навчання
y_pred = neural_network.predict(X)

In [328]:
print("Прогноз до навчання:")
print(y_pred)

Прогноз до навчання:
tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)


In [329]:
# 5. Обчислюємо помилку до навчання
loss = loss_fn(y_pred, y_true)

print("Початкова помилка:")
print(loss)

Початкова помилка:
tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)


In [330]:
# 6. Навчаємо модель
teacher.teach(loss)

In [331]:
# 7. Прогноз після навчання
y_pred = neural_network.predict(X)

print("Прогноз після навчання:")
print(y_pred)

Прогноз після навчання:
tensor([[248.6600],
        [227.9933],
        [376.6600]], device='cuda:0', grad_fn=<AddBackward0>)


In [332]:
# 8.Помилка після навчання
new_loss = loss_fn(y_pred, y_true)

print("Помилка після навчання:")
print(new_loss)

Помилка після навчання:
tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)


Модель стала краще прогнозувати після першого кроку навчання - помилка зменшилася

Початкова помилка:
tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)

Помилка після навчання:
tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)


# Завдання 9

Одного кроку навчання недостатньо, щоб модель навчилася добре прогнозувати ціни. Тому повторіть процес навчання **100 разів**.

### Завдання

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Створіть тензор `X`, який містить дані про квартири.
3. Створіть тензор `y_true`, який містить реальні реальні ціни.
4. Створіть цикл, який виконається **100 разів**.
5. На кожній ітерації циклу:
   * отримайте прогноз моделі `y_pred`;
   * обчисліть помилку за допомогою функції `loss_fn()`;
   * виведіть помилку на екран;
   * передайте помилку в метод `teacher.teach()`, щоб виконати один крок навчання.
    
6. Після завершення циклу ще раз отримайте прогноз моделі.
7. Виведіть:
   * прогнозовані заробітні плати;
   * значення функції втрат.
    
8. Порівняйте отримані прогнози з реальними ціни. Зробіть висновок, чи покращилася точність моделі після 100 кроків навчання.
9. Cпробуйте зменшити навчальний крок в `teacher`

In [333]:
# Створюємо Teacher
teacher = Teacher(neural_network)

In [334]:
# Дані про квартири
X = torch.tensor([
    [40.0, 1.0],
    [60.0, 3.0],
    [100.0, 2.0]
]).to('cuda')

In [335]:
# Реальні ціни
y_true = torch.tensor([
    [210.0],
    [310.0],
    [360.0]
]).to('cuda')

In [336]:
# Навчання 100 разів 

losses = []

for i in range(100):

    y_pred = neural_network.predict(X)

    loss = loss_fn(y_pred, y_true)
    
    print(loss)

    losses.append(loss.item())

    teacher.teach(loss)

print("Початкова помилка:", losses[0])
print("Фінальна помилка:", losses[-1])

tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(56.0045, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8867, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.3023, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9978, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7733, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.9356, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8844, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.2334, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9955, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7711, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.8667, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8822, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.1645, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9933, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7688, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.7979, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8799

In [337]:
# Робимо фінальний прогноз на останній навченій версії
y_pred = neural_network.predict(X)

y_pred

tensor([[249.5928],
        [231.0593],
        [378.6587]], device='cuda:0', grad_fn=<AddBackward0>)

In [338]:
# Фінальна помилку
final_loss = loss_fn(y_pred, y_true)

print("\nПрогнозовані ціни:")
print(y_pred)

print("\nРеальні ціни:")
print(y_true)

print("\nФінальна помилка:")
print(final_loss)


Прогнозовані ціни:
tensor([[249.5928],
        [231.0593],
        [378.6587]], device='cuda:0', grad_fn=<AddBackward0>)

Реальні ціни:
tensor([[210.],
        [310.],
        [360.]], device='cuda:0')

Фінальна помилка:
tensor(45.7307, device='cuda:0', grad_fn=<MeanBackward0>)


In [339]:
# Cпробуйте зменшити навчальний крок в teacher
# Створюємо нову нейронну мережу 
new_neural_network = NeuralNetwork().to('cuda')

In [340]:
# teacher з меншим навчальним кроком
teacher = Teacher(new_neural_network, lr=0.005)

In [341]:
# Дані про квартири
X = torch.tensor([
    [40.0, 1.0],
    [60.0, 3.0],
    [100.0, 2.0]
]).to('cuda')

In [342]:
# Реальні ціни
y_true = torch.tensor([
    [210.0],
    [310.0],
    [360.0]
]).to('cuda')

In [343]:
# Навчання 100 разів 

losses = []

for i in range(100):

    y_pred = new_neural_network.predict(X)

    loss = loss_fn(y_pred, y_true)
    
    print(loss)

    losses.append(loss.item())

    teacher.teach(loss)

print("Початкова помилка:", losses[0])

RuntimeError: X знаходиться на cuda:0, а модель на cpu.

# Завдання 10

У вашому розпорядженні є нейронна мережа, яка прогнозує **ймовірність дощу**.

Модель використовує три ознаки:

* температура повітря (°C);
* вологість повітря (%);
* швидкість вітру (км/год).

Результатом роботи моделі є **два числа**:

1. ймовірність того, що **дощу не буде**;
2. ймовірність того, що **дощ буде**.

Скористайтеся методом `predict()`, щоб визначити ймовірності для таких погодних умов:

* температура — **20°C**;
* вологість — **80%**;
* швидкість вітру — **15 км/год**.



In [ ]:
import torch


class RainNeuralNetwork:
    def __init__(self, device="cpu"):
        self.device = device

        self.W = torch.tensor(
            [
                [0.03, 0.01, 0.02],
                [0.06, -0.01, 0.08]
            ],
            requires_grad=True,
            device=device
        )

        self.b = torch.tensor(
            [-0.2, -0.3],
            requires_grad=True,
            device=device
        )

    def predict(self, X):
        if not isinstance(X, torch.Tensor):
            raise TypeError("X повинен бути об'єктом torch.Tensor.")

        if X.ndim != 2:
            raise ValueError(
                f"Очікується двовимірний тензор, отримано {X.shape}."
            )

        if X.shape[1] != 3:
            raise ValueError(
                f"Очікується 3 ознаки, отримано {X.shape[1]}."
            )

        if not X.is_floating_point():
            raise TypeError(
                "Тензор повинен містити числа з плаваючою комою."
            )

        if X.device != self.W.device:
            raise RuntimeError(
                f"X знаходиться на {X.device}, а модель на {self.W.device}."
            )

        logits = X @ self.W.T + self.b
        return torch.softmax(logits, dim=1)

# Завдання 11

Метеостанція надала прогноз погоди на **4 дні**.

| День | Температура (°C) | Вологість (%) | Швидкість вітру (км/год) |
| ---- | ---------------: | ------------: | -----------------------: |
| 1    |               30 |            40 |                        5 |
| 2    |               22 |            85 |                       18 |
| 3    |               15 |            95 |                       25 |
| 4    |               28 |            60 |                       10 |

Виконайте наступні кроки:

1. Створіть один тензор `X` розміру **(4, 3)**, де кожен рядок містить дані про один день.

2. За допомогою методу `predict()` отримайте ймовірності для кожного дня `y_pred`.

3. Скористайтеся функцією `torch.argmax(..., dim=1)`, щоб визначити прогноз моделі `result`:

   * `0` — дощу не буде;
   * `1` — дощ буде.

4. Виведіть:

   * тензор із ймовірностями `y_pred`;
   * тензор із прогнозами `result`


# Завдання 12

Метеорологи повідомили, чи справді пішов дощ у кожен із чотирьох днів.

| День | Дощ |
| ---- | :-: |
| 1    |  Ні |
| 2    | Так |
| 3    | Так |
| 4    |  Ні |

Для навчання моделі використайте готову функцію втрат **CrossEntropyLoss** з бібліотеки **PyTorch**.

### Завдання

1. Створіть об'єкт функції втрат:

   ```python
   loss_fn = torch.nn.CrossEntropyLoss()
   ```

2. Створіть об'єкт класу `Teacher`, передавши йому модель.

3. Створіть тензор `X`, який містить погодні умови за всі 4 дні.

4. Створіть тензор `y_true`, який містить правильні відповіді:

   * `0` — дощу не було;
   * `1` — дощ був.

5. Створіть цикл, який виконається **100 разів**.

6. На кожній ітерації циклу:

   * отримайте ймовірності від моделі `y_pred`:
   * обчисліть помилку:
   * виведіть помилку
   * передайте помилку методу `teacher.teach(loss)` для виконання одного кроку навчання.

7. Після завершення циклу ще раз отримайте прогноз моделі.

8. За допомогою функції `torch.argmax(..., dim=1)` отримайте прогноз моделі:

   * `0` — дощу не буде;
   * `1` — дощ буде.

9. Виведіть:

   * ймовірності, отримані від моделі;
   * прогнози моделі (`0` або `1`);
   * правильні відповіді
   * значення функції втра


# Завдання 13

Попробуйте натренувати модель використовуючи функцію втрат, яку ви раніше застосовували для задачі визначення цін на квартири (`torch.nn.MSELoss()`), замість `CrossEntropyLoss`.

Метеорологи повідомили, чи справді пішов дощ у кожен із чотирьох днів.

| День | Дощ |
|------|-----|
| 1    | Ні  |
| 2    | Так |
| 3    | Так |
| 4    | Ні  |

**Завдання**

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Створіть тензор `X`, який містить погодні умови за всі 4 дні.
3. Створіть тензор `y_true` розміром (4, 2), де кожен рядок містить інформацію про те чи йшов дощ в цей день:
   - `[1, 0]` — дощу не було
   - `[0, 1]` — дощ був
   -  формат інший тому що ваша функція вимагає щоб `y_true` та `y_pred` були однакового розміру
4. Створіть цикл, який виконається 100 разів.

На кожній ітерації циклу:
- отримайте ймовірності від моделі `y_pred`;
- обчисліть помилку;
- виведіть помилку;
- передайте помилку методу `teacher.teach(loss)` для виконання одного кроку навчання.

6. Після завершення циклу ще раз отримайте прогноз моделі.
7. За допомогою функції `torch.argmax(..., dim=1)` отримайте прогноз моделі:
   - 0 — дощу не буде;
   - 1 — дощ буде.

**Виведіть:**
- ймовірності, отримані від моделі;
- прогнози моделі (0 або 1);
- правильні відповіді;
- значення функції втрат.
